pageserver/src/tenant/timeline.rs

```diff
@@ -3505,12 +3506,27 @@ impl Timeline {
         // min-heap (reserve space for one more element added before eviction)
         let mut heap: BinaryHeap<Hole> = BinaryHeap::with_capacity(max_holes + 1);
         let mut prev: Option<Key> = None;
-        for (next_key, _next_lsn, _size) in itertools::process_results(
-            deltas_to_compact
-                .iter()
-                .map(|l| -> Result<_> { Ok(l.load_keys(ctx)?.into_iter()) }),
-            |iter_iter| iter_iter.kmerge_by(|a, b| a.0 < b.0),
-        )? {
+
+        let mut all_value_refs = Vec::new();
+        for l in deltas_to_compact.iter() {
+            // TODO: replace this with an await once we fully go async
+            all_value_refs.extend(Handle::current().block_on(l.load_val_refs(ctx))?);
+        }
+        // The current stdlib sorting implementation is designed in a way where it is
+        // particularly fast where the slice is made up of sorted sub-ranges.
+        all_value_refs.sort_by_key(|(key, _lsn, _value_ref)| *key);
+
+        let mut all_keys = Vec::new();
+        for l in deltas_to_compact.iter() {
+            // TODO: replace this with an await once we fully go async
+            all_keys.extend(Handle::current().block_on(l.load_keys(ctx))?);
+        }
+        // The current stdlib sorting implementation is designed in a way where it is
+        // particularly fast where the slice is made up of sorted sub-ranges.
+        all_keys.sort_by_key(|(key, _lsn, _size)| *key);
```

Comment:
I wonder(ed on the prev PR) if it would be sensible to just have one vec returned, but unsure how to build up the sizes with that.

Reply:
Yeah that was also my concern when I wondered about that. Ideally we'd have only one vec, yes.